### PR TITLE
Add yihui_style()

### PR DESCRIPTION
**Pull request by Robinlovelace**
A little context: https://github.com/ropensci/software-review-meta/issues/71#issuecomment-444206853

### CODE REVIEW

**Comment by Robinlovelace** [comment]
Hehe I like the `yihuiverse_style()` function name! Seriously though, I agree it's worth thinking about an appropriate name. Originally I called the `equals_style()` but changed it to a more fun name for this PR. I think @yihui should definitely get a say over this. So far I can see/think of 5 options for the style name:

- `equals_style()`
- `eq_assign_style()`
- `equals_assign_style()`
- `yihui_style()`
- `yihuiverse_style()`

I can think of various +s and -s of each name. I'd be happy with any of them and there may be more viable options I've not seen/thought of. I think you guys (primarily @lorenzwalthert and @yihui ) should have a greater say than me over the name so up to you: I'll happily change the name in the PR. Also happy to add tests and make any further changes / put the code elsewhere pending feedback. 

Regarding whether

> it should inherit all other rules from the tidyverse style guide

I'm not sure. Are there any other changes I should make for this to be 'yihuiverse-compliant' :heavy_check_mark:  @yihui hehe?

In summary: thanks for the feedback; I'll make changes pending more feedback, particular relating to the style's name.

**Comment by lorenzwalthert** [comment]
> I'm fine with kidding, but please don't make "Yihui style" or "Yihuiverse" official.

Sure. I think we better find another name.

**Comment by Robinlovelace** [comment]
Thanks for the feedback @yihui I've updated the style name in-line with my original thinking - used `equals_style()` as it's more concise and I think equally understandable as `equals_assign_style()` and more so than `eq_assign_style()`.

**Comment by Robinlovelace** [comment]
Another issue with this PR I've discovered is that the style does not work as expected:

``` r

txt = "x = c(
  1
)"
styler::style_text(txt)
#> x <- c(
#>   1
#> )
styler::style_text(txt, style = styler::equals_style)
#> x = c(
#> 1
#> )
```

<sup>Created on 2018-12-05 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

@lorenzwalthert any suggestions on how to fix this? At present I'm setting the tokens as follows - was thinking of naming the list but not sure if that will help. Here's the function at present in any case:

```r
equals_style <- function() {
 create_style_guide(token = list(force_assignment_op_equals))
}
```

**Comment by Robinlovelace** [comment]
Regarding tests, I've had a first bash but not sure if this fits with your previous tests. Advice on this also appreciated @lorenzwalthert:

``` r
library(styler)
library(testthat)
txt_left_assign = "x <- 1"
txt_equals_assign = "x = 1"
res_equals_style = style_text(txt_left_assign, style = equals_style)
txt_equals_style = as.character(res_equals_style)
test_that("equals_style replaces <- with = in one-liner",{
  expect_identical(
    txt_equals_style,
    txt_equals_assign)
})
```

<sup>Created on 2018-12-05 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by lorenzwalthert** [comment]
As far as tests go, we have a testing framework powered by `test_collection()`. Essentially, there is an *-in.R file and a *-out.R file. The *-in.R file is the input that is transformed and - if it matches the *-out.R file, the test has passed. The files are placed in `tests/testthat` under the category they fit. Please have a look at the documentation for `test_collection()` and see other unit tests. Let me know if there is anything unclear about this.
Also, I hope that you can understand that one or two tests are not sufficient to test the new style guide. There are always unexpected cases such as the one you are presenting above. styler currently has almost 150 tests for the tidyverse style guide and it fooled my over and over again. We need more comprehensive testing. Because you can't think of all cases possible, I think it is helpful to use the source code of another package, style it with the style guide you are suggesting, look at the git diff. Then, pick a few other cases and add them as to the *-in.R test suite. In your case it is pretty easy because everything should be left unchanged for a package that does not contain arrow assignment. We can't add styling of whole packages as unit tests, but we should at least do it once and cover a few cases. That's at least how I did it for [oneliner](https://github.com/lorenzwalthert/oneliner), maybe you have better suggestions. In that case, let me know. At some point, we may want to build infrastructure to do that automatically with travis, but I guess its not a priority now.

As far as the indention problem goes we need `use_raw_indention = TRUE`:

``` r
library(styler)
force_assignment_op_equals <- function(pd) {
  to_replace <- pd$token == "LEFT_ASSIGN"
  pd$token[to_replace] <- "EQ_ASSIGN"
  pd$text[to_replace] <- "="
  pd
}

equals_style <- function() {
  create_style_guide(
    token = list(force_assignment_op_equals),
    use_raw_indention = TRUE
  )
}

txt <- "x = c(
  1
)"
style_text(txt, style = equals_style)
#> x = c(
#>   1
#> )
```

<sup>Created on 2018-12-05 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by lorenzwalthert** [comment]
I also updated `CONTRIBUTING.md` to cover testing. It may contain other interesting information for you about contributing to styler.

**Comment by Robinlovelace** [comment]
Good news on the testing front: `equals_style()` has already proved useful in tidying-up style in a package with equals assignment after commits adding `<-` assignment: https://github.com/ITSLeeds/stats19/commit/327e436bd3e0f3b2e277acd4032e5edc6f24c0ce

Some of those examples could be good in the tests I imagine.

**Comment by lorenzwalthert** [comment]
See my comments in the repo there. Maybe there is an unexpected interaction with #381. Also, you could rebase on master to make the (unrelated) CI failure on R 3.2 go away now.

**Comment by lorenzwalthert** [comment]
See my comments in the repo there. Maybe there is an unexpected interaction with #381. Also, you could rebase on master to make the (unrelated) CI failure on R 3.2 go away now.

**Comment by lorenzwalthert** [comment]
See my comments in the repo there. Maybe there is an unexpected interaction with #381. Also, you could rebase on master to make the (unrelated) CI failure on R 3.2 go away now.

**Comment by lorenzwalthert** [comment]
See my comments in the repo there. Maybe there is an unexpected interaction with #381. Also, you could rebase on master to make the (unrelated) CI failure on R 3.2 go away now.

**Comment by Robinlovelace** [comment]
Think I may have borked the rebase. Can create a new clean PR if that would make sense - suggested way forward unless there are better suggestions.

**Comment by Robinlovelace** [comment]
Think I've borked the rebase. Can create a new clean PR if that would make sense - suggested way forward unless there are better suggestions.
